File: Irrgarten/src/irrgarten/Dice.java
package irrgarten;

/**
 * This class take all the decisions of the game that depend on luck. It will
 * produce a series of random values bounded with some given parameters.
 * @author luisvaldivieso
 */
public class Dice {
    /**
     * Max number of uses of weapons and shields
     */
    private static final int MAX_USES = 5;
    /**
     * Max value for the player/monster intelligence
     */
    private static final float MAX_INTELLIGENCE = 10.0f;
    
    /**
     * Max value for strength of players and monsters
     */
    private static final float MAX_STRENGTH = 10.0f;
    
    /**
     * Probability of resurrection for a player
     */
    private static final float RESURRECT_PROB = 0.3f;
    /**
     * Max number of weapons received when winning a combat
     */
    private static final int WEAPONS_REWARD = 2;
    /**
     * Max number of shields received when winning a combat
     */
    private static final int SHIELDS_REWARD = 3;
    /**
     * Max number of health points received when winning a combat
     */
    private static final int HEALTH_REWARD = 5;
    /**
     * Max attack power of weapons
     */
    private static final int MAX_ATTACK = 3;
    /**
     * Max defensive power of the shields
     */
    private static final int MAX_SHIELD = 2;
    
    /**
     * Random object used to generate a stream of pseudorandom numbers
     */
    private static java.util.Random generator = new java.util.Random();
    
    /**
     * Return a random column or row
     * @param max max value of the columns or rows
     * @return  A random number representing the row or the column, 
     * between [0,max[
     */
    public static int randomPos(int max){
        return generator.nextInt(max);
    }
    /**
     * Returns the index of the player who will start the game
     * @param nPlayers number of current players in the game
     * @return Index of the player who will start the game (between [0,nPlayers[
     */
    public static int whoStarts (int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    /**
     * Gives a random value of intelligence
     * @return A value of intelligence between [0,MAX_INTELLIGENCE[
     */
    public static float randomIntelligence(){
        return generator.nextFloat()*MAX_INTELLIGENCE;
    }
    
    /**
     * Gives a random value of strength 
     * @return A value of strength between [0,MAX_INTELLIGENCE[
     */
    public static float randomStrength(){
        return generator.nextFloat()*MAX_STRENGTH;
    }
    
    /**
     * Determine if a player must be revived based on @param RESURRECT_PROB
     * @return 1(true) if the player must be revived or 0 if not
     */
    public static boolean resurrectPlayer(){
        return generator.nextFloat()<RESURRECT_PROB;
    }
    
    /**
     * Decide how many weapons should the player earn when winning a combat
     * @return Number of weapons given to the player when winning a combat. It
     * is a number between [0,WEAPONS_REWARD]
     */
    public static int weaponsReward(){
        return generator.nextInt(WEAPONS_REWARD+1);
    }
    
    /**
     * Decide how many shields should the player earn when winning a combat
     * @return Number of shields given to the player when winning a combat. It
     * is a number between [0,SHIELDS_REWARD]
     */
    public static int shieldsReward(){
        return generator.nextInt(SHIELDS_REWARD+1);
    }
    /**
     * Decide how many health points will the player earn when winning a combat
     * @return Number of health points given to the player when winning a combat. It
     * is a number between [0,HEALTH_REWARD]
     */
    public static int healthReward(){
        return generator.nextInt(HEALTH_REWARD+1);
    }
    
    /**
     * Return a random weapon power
     * @return A float number Weapon power between [0,MAX_ATTACK[
     */
    public static float weaponPower(){
        return generator.nextFloat()*MAX_ATTACK;
    }
    
    /**
     * Return a random shield power
     * @return A float number Shield power between [0,MAX_SHIELD[
     */
    public static float shieldPower(){
        return generator.nextFloat()*MAX_SHIELD;
    }
    
    /**
     * Generate a random number of uses left
     * @return Number of uses left for a shield or weapon. It is an int number
     * between [0,MAX_USES]
     */
    public static int usesLeft(){
        return generator.nextInt(MAX_USES+1);
    }
    
    /**
     * Generate a random number of competition applied
     * @param competence Max number of competence applied (exclusive)
     * @return A floating point number that represents applied intensity by
     * a weapon between [0,competence[ 
     */
    public static float intensity(float competence){
        return generator.nextFloat()*competence;
    }
    
    /**
     * Determine if a weapon must be discarded. It is easier for a weapon to 
     * be discarded as the @param usesLeft parameter aproaches the constant
     * MAX_USES, and if the uses 
     * @param usesLeft remaining uses of the element
     * @return true if a weapon must be discarded, false otherwise
     */
    public static boolean discardElement(int usesLeft){
        
        if (usesLeft ==MAX_USES) return false;
        else {
            float discardProbability = 1 - (usesLeft/MAX_USES);
            return (generator.nextFloat()<= discardProbability);
        }  
    }
    
    
    
}

File: Irrgarten_ruby/dice.rb
# encoding:utf-8
module Irrgarten

# Class that simulates the dice used in the game. It has methods to generate
# random numbers and booleans, and to decide the next step of a player or
# monster.
class Dice
    
    # Max uses of weapons and shields
    @@MAX_USES = 5
    
    # Max intelligence value for players and monsters
    @@MAX_INTELLIGENCE = 10.0

    # Max value for players and monsters strength
    @@MAX_STRENGTH = 10.0

    # Probability of a player to be resurrected
    @@RESURRECT_PROB = 0.3

    # Max number of weapons received when winning a combat
    @@WEAPONS_REWARD = 2

    # Max number of shield received when winning a combat
    @@SHIELDS_REWARD = 2

    # Max number of health points received when winning a combat
    @@HEALTH_REWARD = 5

    # Max power of weapons
    @@MAX_ATTACK = 3

    # Max protection of shields
    @@MAX_SHIELD = 2

    # Random numbers and booleans generator. In this case the use of a 
    # class variable isnt recommended, because if the generator object is
    # shared across all the methods it could lead to patterns in the 
    # generation of random numbers
    @generator = Random.new

    # @return a random column or row between [0,max[
    def self.random_pos(max)
        @generator.rand(max)
    end

    # @return the index of the player who will start the game, between
    # [0,n_players[
    def self.who_starts(n_players)
        @generator.rand(n_players)
    end

    # @return a value for intelligence between [0,MAX_INTELLIGENCE[
    def self.random_intelligence
        @generator.rand(@@MAX_INTELLIGENCE)
    end

    # @return a random value of strength between [0,MAX_STRENGTH[
    def self.random_strength
        @generator.rand(@@MAX_STRENGTH)
    end

    # @return a boolean representing if the player must be revived.
    # True if the player must be revived or false otherwise.
    def self.resurrect_player 
        (@generator.rand(1.0)<@@RESURRECT_PROB)
    end

    # @return the number of weapons the player should receive when winning
    # a combat, a integer between [0,WEAPONS_REWARD]
    def self.weapons_reward 
        @generator.rand(@@WEAPONS_REWARD+1)
    end 

    # @return the number of shields the player should receive when winning
    # a combat, a integer between [0,SHIELDS_REWARD]
    def self.shields_reward 
        @generator.rand(@@SHIELDS_REWARD+1)
    end  

    # @return the number of health points the player should receive when winning
    # a combat, a integer between [0,HEALTH_REWARD]
    def self.health_reward 
        @generator.rand(@@HEALTH_REWARD+1)
    end  

    # @return a random value between [0, MAX_ATTACK[
    def self.weapon_power
        @generator.rand(@@MAX_ATTACK)
    end

    # Return a random value between [0, MAX_SHIELD[
    def self.shield_power
        @generator.rand(@@MAX_SHIELD)
    end

    # Return the uses a weapon or shield will have, a number between 
    # [0,MAX_USES]
    def self.uses_left
        @generator.rand(@@MAX_USES+1)
    end 

    # Return the competence applied between [0,competence[
    def self.intensity(competence)
        @generator.rand(competence)
    end 

    # Decides if the element must be discarded. The probability of this 
    # event will be greater as uses_left approaches MAX_USES
    def self.discard_element(uses_left)
        discard_probability = 1 - (uses_left / @@MAX_USES)
        @generator.rand < discard_probability
    end

    # Decides the next step of a player or monster. If the random
    # number generated is less than the intelligence, the player will move
    # in the desired direction. Otherwise, the player will move in a random
    # valid direction.
    # @param preference [String] the desired direction of the player
    # @param valid_moves [Array<String>] the valid directions the player can move
    # @param intelligence [Float] the intelligence of the player
    # @return [String] the next step of the player
    def self.next_step(preference, valid_moves, intelligence)
        # Probability of moving in the desire direction
        probability = @generator.rand*@@MAX_INTELLIGENCE
        if probability < intelligence
            preference
        else
            # Generate a random valid move
            random_index_direction = @generator.rand(valid_moves.length)
            valid_moves[random_index_direction]
        end
    end
end

end
